internal/blobstore/s3store.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package blobstore // import "gopkg.in/juju/charmstore.v5-unstable/internal/blobstore"
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"log"
+
+	"gopkg.in/amz.v3/aws"
+	"gopkg.in/amz.v3/s3"
+	"gopkg.in/errgo.v1"
+)
+
+type s3Store struct {
+	bucket string
+	getS3  func() *s3.S3
+}
+
+// NewS3 createa a new S3 backed blobstore Store
+func NewS3(pc *ProviderConfig) *Store {
+	return &Store{newS3(pc)}
+}
+
+func newS3(pc *ProviderConfig) *s3Store {
+	getter := getS3(pc)
+	s := &s3Store{
+		bucket: pc.BucketName,
+		getS3:  getter,
+	}
+	s.createBucket()
+	return s
+}
+
+func (s *s3Store) createBucket() {
+	bucket, err := s.getS3().Bucket(s.bucket)
+	if err != nil { //this only happens on invalid bucket name that is when a name has any of /:@
+		panic(err)
+	}
+	err = bucket.PutBucket(s3.Private)
+	if err != nil {
+		log.Println("Failed to create bucket", err)
+	}
+}
+
+func (s *s3Store) Put(r io.Reader, name string, size int64, hash string, proof *ContentChallengeResponse) (_ *ContentChallenge, err error) {
+	err = s.PutUnchallenged(r, name, size, hash)
+	return
+}
+
+func (s *s3Store) PutUnchallenged(r io.Reader, name string, size int64, hash string) error {
+	svc := s.getS3()
+	bucket, _ := svc.Bucket(s.bucket) // Ignoring the error because we know this bucket name is valid.
+	err := bucket.PutReader(name, r, size, "application/octet-stream", s3.Private)
+	if err != nil {
+		logger.Errorf("put failed :%s", err)
+		return errgo.Mask(err)
+	}
+	logger.Debugf("successful put %s in bucket %s", name, s.bucket)
+	return nil
+}
+
+func (s *s3Store) Open(name string) (ReadSeekCloser, int64, error) {
```

Comment:
I can see you're not happy with this implementation, and neither am I. perhaps you can use List up front to get the size, and then wrap rc in a type that provides a limited in memory data cache so that you can seek back to the start if necessary (which is the only sensible place to seek). There might already be such a thing implemented somewhere.